### DISCUSSION TITLE
When will Fooocus for Mac come out?

### DISCUSSION THREAD

**Discussion by Latirin** [lllyasviel/Fooocus#388] [General]
I really neeeeeeed that!

**Comment by lsooxlla8** [comment]
And are there any plans for m1?

**Comment by barepixels** [comment]
someone managed to get it to work

https://github.com/lllyasviel/Fooocus/pull/129

UPDATE:  Use this fork  https://github.com/jorge-campo/Fooocus/tree/main

**Reply by AlexUrrutia** [reply]
404

**Comment by Acephalia** [comment]
Would probably be much faster when Core ML is utilised. Both Draw Things and Diffusers are reasonably fast now. Here's hoping we get an official version soon.

**Comment by barepixels** [comment]
I have an M1 Macbook Pro but never bother to do Stable Diffusion or Mining (no longer) with it.  I built a PC for those tasks.  The bottom line use the right tool for the job.  Blame Apple for not being friendly with NVidia

**Comment by vesper8** [comment]
You do not need to do anything special to make it work on Apple Silicon Mac.. I followed the Linux instructions and it worked like a charm on first try.

**Reply by Latirin** [reply]
thanks for reply, I installed successfully, but when I generate, it got a "RuntimeError: User specified autocast device_type must be 'cuda' or 'cpu'", do you know why is that and how to fix it?

**Reply by parrydigm** [reply]
I managed to get it working using homebrew to install miniconda. However, this is the slowest way I've found yet to run Stable Diffusion 🤣 man, it is slllllow. It might be better to try using chatgpt as a prompt generator and using draw things instead until we get something optimised.

**Comment by eddyizm** [comment]
@vesper8 Agree - it was relatively easy to get it working. Not the fastest but not the slowest and there's still some quirks but honestly, having tried it on 4 different platforms, they all have quirks.

**Comment by KittenYang** [comment]
On my M2 Pro Mac, why it's so slow? 
`epoch | 2/30 [07:05<1:39:39, 213.55s/it]`

**Reply by barepixels** [reply]
Apple and Nvidia don't get along and Nvidia own CUDA a technology that Stable Diffusion use.  Blame Apple for not using Nvidia graphic cards

**Reply by parrydigm** [reply]
Apple and nVidia's relationship doesn't stop AUTOMATIC1111 being _orders of _magnitude_ faster on macOS than Fooocus. It also doesn't stop Draw Things from being faster than both. Sure, running SD on Windows with an nVidia card is faster than on macOS, but Fooocus is a real outlier here.

**Comment by lgwjames** [comment]
try a amd mac im on, 15-20 mins a image

**Comment by AlexUrrutia** [comment]
Testing on MacBook Pro M3 Pro
<img width="1496" alt="Screenshot 2024-03-18 at 9 43 18 PM" src="https://github.com/lllyasviel/Fooocus/assets/7042361/f6b0d188-7e49-470e-be7b-e4c399eddb99">

**Reply by AlexUrrutia** [reply]
**Total time: 1853.52 seconds**
![image](https://github.com/lllyasviel/Fooocus/assets/7042361/156a8887-1910-45f6-98b2-426e79f339b7)

**Reply by KittenYang** [reply]
Way slower than SD WebUI. No idea why

**Comment by tresorama** [comment]
Anyone tried running this in a Linux container with Docker on Apple M1/M2/M3? 
Same slow performance ?

**Reply by barepixels** [reply]
the keyword is CUDA which is exclusively Nvidia.  Part of the reason why Nvidia is bigger than Amazon

**Reply by arkpatrick** [reply]
When I try to start the service using docker, I get the following error. I'm running this on an Apple M3 Max with 48G memory running Sonoma 14.4.1. 

Fooocus % docker compose up
Attaching to app-1
Gracefully stopping... (press Ctrl+C again to force)
Error response from daemon: could not select device driver "nvidia" with capabilities: [[compute utility]]
Fooocus %

**Reply by mashb1t** [reply]
Apple Silicone doesn't allow hardware acceleration using Docker. Please use non-Docker Fooocus for Apple hardware.

**Comment by bazmatic** [comment]
I'm running Fooocus on a Macbook Pro M2. It's working pretty well! I didn't have to change anything, just started generating images straight out of the box. Great work!

**Reply by tresorama** [reply]
Was slow before as other [reported](https://github.com/lllyasviel/Fooocus/discussions/388#discussioncomment-8835096) ?

**Reply by mashb1t** [reply]
@tresorama I personally don't have any issues on a M1 Macbook, image generation takes very roughly 90-100s for performance speed.

**Reply by tresorama** [reply]
Thanks @mashb1t .  Great to hear that, it's a really decent time!
In my few tests I've done in Google Colab, 30seconds is aligned to generating a 1152 × 896 image.

I need to upgrade my old Macbook Pro i7 so I'm opting for Apple M-Family.
Could you share the exact Macbook model that you use ?

**Reply by mashb1t** [reply]
@tresorama I'm deeply sorry, I double-checked and was mistaken, it takes about roughly 90-100 instead of the mentioned and now edited 30-40 seconds to generate an image on the official Fooocus main branch.
```
100%|██████████| 30/30 [01:36<00:00,  3.20s/it]
100%|██████████| 30/30 [01:40<00:00,  3.35s/it]
```

Still can't reproduce the MPS issue though (M1 Max Macbook Pro 16 inch, 2021, 32GB RAM).
Please don't base your decision solely on the performance of Fooocus or my numbers and do your own research if the device fits your needs. Hope this helps and sorry again.

**Comment by parrydigm** [comment]
> @tresorama I'm deeply sorry, I double-checked and was mistaken, it takes about roughly 90-100 instead of the mentioned and now edited 30-40 seconds to generate an image on the official Fooocus main branch.

Even 90-100 would be a vast improvement to what I see. This is Fooocus main branch on an M2 mini 16GB:

10%|████▎                                      | 3/30 [05:13<45:59, 102.20s/it]

...which is a bit slower than what you see 🤣 I do note mine produces an error:

anisotropic.py:132: UserWarning: The operator 'aten::std_mean.correction' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:13.)

I would guess this is the heart of the problem - mine is falling back to CPU and yours is *likely* using Apple Silicon acceleration. Do you have any guidance on your installation process? I'd like to try what you did (nuke my install, start again) so if you could throw some advice my way I'll try replicating.

**Reply by mashb1t** [reply]
@parrydigm I've used the unofficial guideline for Mac you can find in the Fooocus readme: https://github.com/lllyasviel/Fooocus?tab=readme-ov-file#mac

**Comment by parrydigm** [comment]
> @parrydigm I've used the unofficial guideline for Mac you can find in the Fooocus readme: https://github.com/lllyasviel/Fooocus?tab=readme-ov-file#mac

Thanks, that's what I used too.

I've just done a fresh install on my M1 MacBook Air (16GB) to see if something weird is going on with my M2 mini. Followed the same instructions, used the nightly build per the Metal Pytorch instructions (https://developer.apple.com/metal/pytorch/ with the output correctly showing "tensor([1.], device='mps:0')"), and... I still get the MPS backend CPU fallback warning, and incredibly slow render speed. So, using the instructions, but getting a hard fail on Metal use.

Looking at the CPU load on the unit, it's more or less just idling - 85% idle while rendering, but has McMassive memory pressure (Python using 25GB of memory, and of course that means we're at about 18GB of swap used). So, there's something going on where a) we're not using Metal acceleration, and b) we're using stupendous amounts of memory (AUTOMATIC1111, Draw Things do not have this behaviour).

I'm at a loss; my hunch is there's something busted in the nightly pytorch used on my machinery 🤷🏼 Another possible option is something iffy with homebrew on my machine (which I also run) - totally at a loss as to how to determine if that's the case tho.

**Comment by parrydigm** [comment]
> I'm at a loss; my hunch is there's something busted in the nightly pytorch used on my machinery 🤷🏼 Another possible option is something iffy with homebrew on my machine (which I also run) - totally at a loss as to how to determine if that's the case tho.

Hm, okay - I completely removed Homebrew, which caused Fooocus a few issues so it was def. pointed there. But after a scrub and re-install (it's now unable to use Homebrew and must rely on the pytorch nightly) it still doesn't work 🙁 

```
anisotropic.py:132: UserWarning: The operator 'aten::std_mean.correction' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:13.)
  s, m = torch.std_mean(g, dim=(1, 2, 3), keepdim=True)
 60%|█████████████████████████▏                | 18/30 [37:28<24:50, 124.22s/it]
```

I dunno, this feels like it should just work, but isn't, implying a bug somewhere. ChatGPT 🤣 says the error indicates an issue with the MPS backend at fault, but it's also told me that pasta is higher in protein than carbs, so mileage varies. The error itself points to Fooocus's anisotropic.py module using an unsupported function, so I imagine there's a capability hole here.

```
ChatGPT: It's a bit tricky to pinpoint the exact culprit without more context, but typically, when you see a warning like this, it indicates that there might be an issue with the compatibility between the code (in this case, anisotropic.py) and the backend it's running on (MPS in this case).

It could be a case where the MPS backend doesn't fully support certain operations that the code is trying to perform, leading to the fallback to CPU execution. However, it's also possible that the code could be optimized or updated to better leverage the capabilities of the MPS backend.

I'd recommend investigating further by checking the documentation for both the MPS backend and the code in anisotropic.py to see if there are any known compatibility issues or optimizations that can be made. Additionally, checking for any updates or fixes for either the backend or the code could also help resolve the issue.
```

**Reply by IPv6** [reply]
BTW, i. have similar problems with FreeU enabled. FreeU on Mac mini forces CPU execution too.

**Comment by parrydigm** [comment]
> BTW, i. have similar problems with FreeU enabled. FreeU on Mac mini forces CPU execution too.

Ah okay - so it could be something of a wider use scenario that the code is asking the MPS backend for that's not implemented. This still doesn't help us mortals 🤣 get a solution to Fooocus on Mac, but it may highlight that Apple needs to do more work, or the devs need to recognise the gap and implement a workaround.

**Comment by tresorama** [comment]
Found this blog post 
https://appleinsider.com/articles/23/12/13/apple-silicon-m3-pro-blows-away-nvidia-rtx-4090-gpu-in-ai-benchmark

The article compares Apple Silicon and Nvidia 4090 in an AI test (not using a Stable Diffusion task, but a Whisper by Open AI task).
There is a mention to a `MLX` framework.
Sneak peaking into mlx repo, they have a stable-diffusion demo.

https://github.com/ml-explore
https://github.com/ml-explore/mlx-examples/tree/main/stable_diffusion

---
Now i have few questions...

- This `mlx` framework can be seen as the python API that software can consume to talk to the Apple GPU Hardware ? Same as cuda in Nvidia world ??  
- If yes, this `mlx` can be used as `cuda` replacement inside python app that use Stable Diffusion?
- Can be used in Fooocus to have better performance when machine is Apple Silicon ?? 

--- 
I apologize in advance for my ignorance here and potential silly questions, AI/ML/GPU is not my thing...
